add auto_release

tachella opened the pull request:
add automatic releases every 21 days, which add a patch to the current version ie going from `a.b.c` to `a.b.c'` where `c'=c+1`. The github action should check first with TestPyPI that everything is working fine before doing the final release.

### Checks to be done before submitting your PR
- [ ] `python3 -m pytest tests/` runs successfully.
- [ ] `black .` runs successfully.
- [ ] `make html` runs successfully (in the `docs/` directory).
- [ ] Updated docstrings related to the changes (as applicable).
- [ ] Added an entry to the [CHANGELOG.rst](../CHANGELOG.rst).

Andrewwango commented on docs/source/conf.py:
```diff
@@ -19,15 +19,27 @@
+project = metadata["name"]
+copyright = "deepinverse contributors"
+
+author = ""
+for auth in metadata["authors"]:
+    if author:
+        author += ", "
+    author += auth["name"]
```
```suggestion
author = ", ".join(auth["name"] for auth in metadata["authors"])
```

Andrewwango commented on docs/source/conf.py:
```diff
@@ -19,15 +19,27 @@
-    "Julian Tachella, Matthieu Terris, Samuel Hurault, Dongdong Chen and Andrew Wang"
-)
-release = "0.3"
+try:
```
imports to top of file

Andrewwango commented on deepinv/__about__.py:
```diff
@@ -7,11 +15,14 @@
+__title__ = metadata["name"]
+__summary__ = metadata["description"]
+__version__ = metadata["version"]
+authors = ""
+for author in metadata["authors"]:
+    if authors:
+        authors += ", "
+    authors += author["name"]
+__author__ = authors
```
```suggestion
__author__ = ", ".join(auth["name"] for auth in metadata["authors"])
```

Andrewwango commented on .github/workflows/release.yml:
```diff
@@ -0,0 +1,82 @@
+        with:
+          tag_name: v${{ steps.bump.outputs.version }}
+          name: Release v${{ steps.bump.outputs.version }}
+          files: dist/*
```
Can we add a description to the release e.g. saying the latest features integrated since the last release (i.e. automating what you do already?)

tachella replied:
done, the script should read the changes from the CHANGELOG

Andrewwango replied:
nice

Andrewwango commented on .github/workflows/release.yml:
```diff
@@ -0,0 +1,82 @@
+    name: Auto version, test & publish
+    runs-on: ubuntu-latest
+
+    steps:
```
Since I can't run these now to verify that they work, can I just check - are these steps below the exact commands you use currently to manually release?

tachella replied:
yes, these are the typical steps that I do to release

Andrewwango reviewed (changes requested):
Thanks Julian, some comments:

Regarding `pyproject.toml`:
```python
authors = [{name="Julian Tachella", email="tachellajulian@gmail.com"}]
description = "Pytorch library for solving inverse problems with deep learning"
```
1. Do you think we should include maintainers'/paper authors' authorship here or just you?
2. Description should better match index: **"DeepInverse: a PyTorch library for imaging with deep learning"**

tachella commented on .github/scripts/bump_patch_version.py:
```diff
@@ -0,0 +1,56 @@
+
+# Find and update the version line
+for i, line in enumerate(lines):
+    match = version_pattern.match(line)
```
it seems that tomllib which is the standard python parser (from 3.11) does not have a writing function. The toml library does, but it is non-standard and would add another dependency to the library.

tachella commented on .github/scripts/bump_patch_version.py:
```diff
@@ -0,0 +1,56 @@
+with open(path, "r", newline="") as f:
+    lines = f.readlines()
+
+# More precise pattern that matches the whole version line
```
see below regarding the tomllib library

tachella commented:
> Thanks Julian, some comments:
> 
> Regarding `pyproject.toml`:
> 
> ```python
> authors = [{name="Julian Tachella", email="tachellajulian@gmail.com"}]
> description = "Pytorch library for solving inverse problems with deep learning"
> ```
> 
> 1. Do you think we should include maintainers'/paper authors' authorship here or just you?
> 2. Description should better match index: **"DeepInverse: a PyTorch library for imaging with deep learning"**

1. I've changed to DeepInverse Maintainers and a link to the maintainer page to avoid having multiple sources of the same info
2. fixed

Andrewwango commented on .github/workflows/release.yml:
```diff
@@ -74,9 +92,24 @@ jobs:
+      - name: Extract changelog description
+        id: changelog
+        run: |
+          python .github/scripts/get_latest_changelog.py > changelog.txt
```
your script saves a changelog.txt and prints to stdout just a message. So this line should be just run the script, then read changelog.txt (making sure to read from .github/scripts/changelog.txt

Andrewwango commented on .github/scripts/get_latest_changelog.py:
```diff
@@ -0,0 +1,89 @@
+    body = match.group("body").rstrip()
+
+    # Extract current version from pyproject.toml or bump script
+    version = get_version_from_toml("../../pyproject.toml")
```
you need to then increment the version

Andrewwango commented on docs/source/conf.py:
```diff
@@ -12,10 +12,7 @@
-    import tomllib
-except ImportError:
-    import tomli as tomllib
+import tomlkit
```
nice, but __about__ should also use this now

tachella replied:
fixed

Andrewwango commented on .github/scripts/bump_version.py:
```diff
@@ -0,0 +1,58 @@
+
+
+def increment_version(version, increment="patch"):
+    """Increment the patch version number (last digit)."""
```
change this docstring

Andrewwango reviewed (changes requested):
Please bump changelog too

Andrewwango commented on .github/scripts/get_latest_changelog.py:
```diff
@@ -81,14 +84,16 @@ def update_changelog(changelog_path):
+    increment = sys.argv[2] if len(sys.argv) > 2 else "patch"
 
-    update_changelog(sys.argv[1])
+    update_changelog(sys.argv[1], increment)
```
in release.yml pass in increment specified in workflow

tachella commented on .github/scripts/get_latest_changelog.py:
```diff
@@ -81,14 +84,16 @@ def update_changelog(changelog_path):
-    if len(sys.argv) != 2:
-        print("Usage: python get_latest_changelog.py path/to/CHANGELOG.rst")
+    if len(sys.argv) not in [2, 3]:
+        print("Usage: python bump_patch_version.py path/to/pyproject.toml patch")
```
you can use as

`python bump_patch_version.py path/to/pyproject.toml`

`python bump_patch_version.py path/to/pyproject.toml patch`

Andrewwango replied:
but this is get_latest_changelog

tachella commented on deepinv/__about__.py:
```diff
@@ -1,10 +1,7 @@
-with open("pyproject.toml", "rb") as f:
-    metadata = tomllib.load(f)["project"]
+with open("../pyproject.toml", "rb") as f:
+    metadata = tomlkit.load(f)["project"]
```
true, thanks for spotting this -after so many changes of library I got confused

tachella commented on .github/workflows/release.yml:
```diff
@@ -68,7 +68,7 @@ jobs:
         id: bump
         run: |
-          python .github/scripts/bump_version.py pyproject.toml ${{github.event.inputs.bump_type}}$
+          python .github/scripts/bump_version.py pyproject.toml ${{github.event.inputs.bump_type}}
```
what do you mean?

tachella commented on .github/workflows/release.yml:
```diff
@@ -120,6 +120,6 @@ jobs:
           git config user.name "github-actions"
           git config user.email "github-actions@github.com"
-          git add pyproject.toml CHANGELOG.rst
+          git add pyproject.toml CHANGELOG.rst ${{github.event.inputs.bump_type}}
```
yes, I had added it in the wrong place - now it is fixed

tachella commented on .github/scripts/get_latest_changelog.py:
```diff
@@ -0,0 +1,99 @@
+
+if __name__ == "__main__":
+    if len(sys.argv) not in [2, 3]:
+        print("Usage: python get_latest_changelog.py path/to/CHANGELOG.md patch")
```
done

mh-nguyen712 commented on pyproject.toml:
```diff
@@ -32,6 +32,7 @@ dependencies = [
     "wandb",
     "scipy",
     "natsort",
+    "tomlkit",
```
Do we really need this dependency here? This line https://github.com/deepinv/deepinv/pull/573/files#diff-87db21a973eed4fef5f32b267aa60fcee5cbdf03c67fafdc2a9b553bb0b15f34R66 is not enough?
